// pomelo/lib/connectors/hybridconnector.js
var net = require('net');
var tls = require('tls');
var util = require('util');
var EventEmitter = require('events').EventEmitter;

var HybridSocket = require('./hybridsocket');
var Switcher = require('./hybrid/switcher');
var Handshake = require('./commands/handshake');
var Heartbeat = require('./commands/heartbeat');
var Kick = require('./commands/kick');
var coder = require('./common/coder');

var curId = 1;

/**
 * Connector that manager low level connection and protocol bewteen server and client.
 * Develper can provide their own connector to switch the low level prototol, such as tcp or probuf.
 */
var Connector = function(port, host, opts) {
  if (!(this instanceof Connector)) {
    return new Connector(port, host, opts);
  }

  EventEmitter.call(this);

  this.opts = opts || {};
  this.port = port;                       //端口
  this.host = host;                       //ip
  this.useDict = opts.useDict;            //是否使用用户字典
  this.useProtobuf = opts.useProtobuf;    //是否使用protobuf
  this.handshake = new Handshake(opts);   //握手对象commands/handshake.js
  this.heartbeat = new Heartbeat(opts);   //心跳对象commands/heartbeat.js
  this.distinctHost = opts.distinctHost;  //是否区分本地host
  this.ssl = opts.ssl;                    //是否为ssl

  this.switcher = null;
};

util.inherits(Connector, EventEmitter);

module.exports = Connector;

/**
 * Start connector to listen the specified port
 */
Connector.prototype.start = function(cb) {
  var app = require('../pomelo').app;
  var self = this;

  var gensocket = function(socket) {
    var hybridsocket = new HybridSocket(curId++, socket);  //关联HybridSocket.js
    hybridsocket.on('handshake', self.handshake.handle.bind(self.handshake, hybridsocket));
    hybridsocket.on('heartbeat', self.heartbeat.handle.bind(self.heartbeat, hybridsocket));
    hybridsocket.on('disconnect', self.heartbeat.clear.bind(self.heartbeat, hybridsocket.id));
    hybridsocket.on('closing', Kick.handle.bind(null, hybridsocket));
    self.emit('connection', hybridsocket);   //触connector的connection事件
  };

  this.connector = app.components.__connector__.connector;
  this.dictionary = app.components.__dictionary__;
  this.protobuf = app.components.__protobuf__;
  this.decodeIO_protobuf = app.components.__decodeIO__protobuf__;

  if(!this.ssl) {
    this.listeningServer = net.createServer();             //关联tcp服务，hybrid/switcher.js
  } else {
    this.listeningServer = tls.createServer(this.ssl);
  }
  this.switcher = new Switcher(this.listeningServer, self.opts);

  //此处的sokcet是由wsprocessor或者tcpprocessor中创建的新的客户端的socket
  this.switcher.on('connection', function(socket) {
    gensocket(socket);
  });

  if(!!this.distinctHost) {
    this.listeningServer.listen(this.port, this.host);
  } else {
    this.listeningServer.listen(this.port);
  }

  process.nextTick(cb);
};

Connector.prototype.stop = function(force, cb) {
  this.switcher.close();
  this.listeningServer.close();

  process.nextTick(cb);
};

Connector.decode = Connector.prototype.decode = coder.decode;

Connector.encode = Connector.prototype.encode = coder.encode;
