// pomelo-base-plugin/lib/const/sysConst.js
/**
 *----------------------------------------------
 * Author: lenon
 * Create Time: 2016/3/2 10:38
 * Desc: (系统常量定义)
 * Changed list:
 *        Change author:(若修改请指明修改者)
 *        Change desc:(修改描叙)
 *        Change time:(修改的时间)
 *----------------------------------------------
 **/

module.exports = {
    DEFAULT_SETTINGS : {
        RANDOM_SEED_LENGTH : 6,                     // 随机种子长度;
        SERVICES_UPDATE_INTERVAL : 1000,            // 每帧毫秒数
        SERVICES_DIR_NAME : 'services',             // 服务目录名
        MEMORY_WATCH_STATUS : false,                // 内存监控状态，false为关闭,true为开启
        MEMORY_GC_INTERVAL : 300000                 // 内存回收间隔
    },
    LIFECYCLE_FUNCTION : {                          // pomelo进程的生命周期函数，详细见pomelo的lifecycle函数
        BEFORE_STARTUP : 'beforeStartup',           // 进程启动前
        BEFORE_SHUTDOWN : 'beforeShutdown',         // 进程启动后
        AFTER_STARTUP : 'afterStartup',             // 进程启动后
        AFTER_START_ALL : 'afterStartAll'           // 所有进程启动完毕
    },
    SERVICES_BUILT_IN_FUNCTION : {                  // 服务内置函数
        STARTUP : 'startup',                        // 启动
        UPDATE : 'update',                          // 更新
        SHUTDOWN : 'shutdown'                       // 关闭
    }
};
